### ISSUE TITLE
Are these schemas equivalent?

### ISSUE THREAD

**Issue by opethe1st** [json-schema-org/json-schema-spec#859]
Tried to check through issues to see if this question has been answered before but I couldn't see anything.

Is this schema (1)
```json
{
 "$ref" : {"#/$defs/object"},
 "additionalProperties" : false
 "$defs" : {
     "object": {
        "properties": {
            "data": true,
         }
     }
}
```
equivalent to this (2)?
```json
{
 "properties: {
     "data": true,
  }
 "additionalProperties" : false
}
```
Since $ref now alongs sibling keywords in the latest specification.

A related question is this. 
For  allOf.
is this schema (3)
```json
{
  "allOf": [
     {"properties": {"data": true},
     {"additionalProperties": false}
   ]
}
```

equivalent to this (4)
```json 
{
  "properties": {"data": true},
  "additionalProperties": false
}
```

The reason I am asking is that I thought that they wouldn't be since they are in different schemas and so don't interact. According to the spec, additionalProperties has no effect if there is no properties keyword side by side.

**Comment by Relequestual** [comment]
They are not equivalent.

As you identified in the second part of your question, they are not in the same schema object.

In terms of the first part of your question, the same applies. We made it clearer that '$ref' is an applicator keyword and is not a transclusion operation.

We did add 'unevaluatedProperties' which works similar but based on the annotation results of adjacent schema objects.

It would work the same in both of your examples. 

I would give more examples but I’m limited to mobile right now. 

Let me know if you have any further questions or feel free to close the issue.

**Comment by opethe1st** [comment]
Thanks for taking time to respond @Relequestual. 
I definitely need more examples to illustrate how all of this works.

This question was inspired by this example in the specification.
```
{
    "$schema": "https://json-schema.org/draft/2019-09/schema",
    "$id": "https://example.com/tree",
    "$recursiveAnchor": true,
    "type": "object",
    "properties": {
        "data": true,
        "children": {}
    }
}
```
```
{
    "$schema": "https://json-schema.org/draft/2019-09/schema",
    "$id": "https://example.com/strict-tree",
    "$recursiveAnchor": true,
    "$ref": "tree",
    "unevaluatedProperties": false
}
```
// instance with misspelled field
```
{
    "children": [ { "daat": 1 } ]
}

```
the properties in "https://example.com/tree", is not considered when evaluating the unevaluatedProperties in "https://example.com/strict-tree" right?  (because they are not adjacent?). And so this example doesn't work, since it only applies when properties, additionalProperties and patternProperties are present right? 

I think it would be a good idea to make explicit that dependent keywords don't depend on each other across schema boundaries. Particularly when used with `$ref` or `$recursiveRef`.
So you will be able to do something like

``json
{
    "$schema": "https://json-schema.org/draft/2019-09/schema",
    "$id": "https://example.com/max10",
    "maxLength": 10
}
```
```
{
    "$schema": "https://json-schema.org/draft/2019-09/schema",
    "$id": "https://example.com/min-0-max-10",
    "$ref": "max10",
    "minLength": 4
}
```
and it will work like expected to constrain the instance when it is a string to be of length 4 to 10.
but if you do this with properties that depend on each other, they wont interact - (like the additionalProperties and properties example above.)


Except that is not what is intended by the specification.

**Comment by Relequestual** [comment]
Sorry for the delay in getting back to you...

> the properties in "https://example.com/tree", is not considered when evaluating the unevaluatedProperties in "https://example.com/strict-tree" right? (because they are not adjacent?).

I'm afraid you missunderstand. This is exactly the point of unevaulatedProperties, that it can "see through" applicator keywords (like `$ref`) in order to consider `properties` in non adjacent but applicable schemas.

It's a little more complex in the example from the spec appendix because that section is really about how the recursive anchor works.

You may want to take a look at the example I used in a talk last year: https://speakerdeck.com/relequestual/json-schema-draft-8-to-vocabularies-and-beyond?slide=23

![Screen Shot 2020-02-21 at 12 40 28](https://user-images.githubusercontent.com/731158/75035442-5f198100-54a7-11ea-8bd9-2068bb128fb6.png)

As you have identified, `additionalProperties` is dependent on keywords in the SAME schema object.

We do actually spell this out as part of the 2019-09 core document: https://tools.ietf.org/html/draft-handrews-json-schema-02#section-9.1


> 9.1. Keyword Independence
> 
> Schema keywords typically operate independently, without affecting each other's outcomes.
> 
> For schema author convenience, there are some exceptions among the keywords in this vocabulary:
> 
> - "additionalProperties", whose behavior is defined in terms of "properties" and "patternProperties"
> - "unevaluatedProperties", whose behavior is defined in terms of annotations from "properties", "patternProperties", "additionalProperties" and itself
> - "additionalItems", whose behavior is defined in terms of "items"
> - "unevaluatedItems", whose behavior is defined in terms of annotations from "items", "additionalItems" and itself


When we talk about annotation collection, these are as a result of other keywords and schemas being applied to the instance, and you can read the requirements for annotation collection for each keyword, and in an overall section.

Essentially, `unevaluatedProperties` can "see through" applicator keywords, in layman's terms.

If it coudln't, then it would function the same as `additionalProperties`!

**Comment by opethe1st** [comment]
Thanks for the very detailed reply 😄 . 
I suspect this is going to be hard to implement but thanks to the cop out in the specification. I am going to skip supporting unevaluated* keywords.

**Comment by handrews** [comment]
@opethe1st in the forthcoming 2020-03 (or at the latest 2020-04) draft we are putting `unevaluatedProperties` and `unevaluatedItems` in their own vocabulary, so your implementation would be able to tell whether a schema (based on its meta-schema) requires support for those keywords.  See #853 for details.

The current cop-out is a little too unpredictable in behavior.  2020-03 will be a clarification of 2019-09, so essentially all work you do on 2019-09 will roll over to 2020-03 without you needing to do anything.  We will strongly encourage implementations to just roll forward to 2020-03 since *complete* 2019-09 implementations are very rare (I think just one) and 2020-03 will align with OpenAPI 3.1.

**Comment by handrews** [comment]
@opethe1st is there more to address here or can we close this out?

**Comment by opethe1st** [comment]
Nothing else to address at the moment. I will close this.
